core/server/common/src/main/java/alluxio/master/MasterRegistry.java

```diff
@@ -116,9 +95,56 @@ public MasterRegistry() {}
   /**
    * @return a collection of all the registered masters
    */
-  public synchronized Collection<Master> getMasters() {
-    List<Master> masters = new ArrayList<>(mRegistry.values());
-    Collections.sort(masters, COMPARATOR);
+  public Collection<Master> getMasters() {
+    List<Map.Entry<Class<?>, Master>> entries = new ArrayList<>(mRegistry.entrySet());
+    Collections.sort(entries, new DependencyComparator());
+    List<Master> masters = new ArrayList<>();
+    for (Map.Entry<Class<?>, Master> entry : entries) {
+      masters.add(entry.getValue());
+    }
     return masters;
   }
+
+  private Set<Class<?>> getTransitiveDeps(Class<?> clazz) {
+    Set<Class<?>> result = new HashSet<>();
+    Set<Class<?>> visited = new HashSet<>();
+    Deque<Class<?>> queue = new ArrayDeque<>();
+    queue.add(clazz);
+    while (!queue.isEmpty()) {
```

Comment:
The number of masters is limited, so I'd prefer to simplify the DFS to not do caching or use separate result and visited sets, e.g.
```java
Set<Master> deps = new HashSet<>();
Deque<Master> queue = new ArrayDeque<>();
queue.add(master)
while (!queue.isEmpty()) {
  for (Master dep : queue.pop().getDependencies()) {
    if (dep == master) {
      throw new RuntimeException(ExceptionMessage.DEPENDENCY_CYCLE.getMessage());
    }
    if (deps.contains(dep)) {
      continue;
    }
    queue.add(dep);
    deps.add(dep);
}
return deps
```